EL7: exception when parsing version string

```
# ncm-ncd --configure --all
 
[INFO] NCM-NCD version 16.8.0 started by root at: Tue Sep 20 12:11:54 2016
updated /var/lib/ccm/profile.3/ccm-active-profile.3-20593
[INFO] No components specified, getting all active ones.
[ERROR] component afsclt has invalid version from config 16.6.0 (bug in profile?): Invalid version format (non-numeric data) at /usr/lib/perl/NCD/ComponentProxy.pm line 267.
 
[ERROR] Failed to create ComponentProxy for component afsclt
[ERROR] No components to dispatch.
[ERROR] Uncaught exception!
[ERROR] Calling stack is:
                LC::Exception::throw_error called at /usr/lib/perl/CAF/Object.pm line 106
 
                CAF::Object::new called at /usr/lib/perl/NCD/ComponentProxyList.pm line 661
 
                NCD::ComponentProxyList::get_proxies called at /usr/lib/perl/NCD/ComponentProxyList.pm line 745
 
                NCD::ComponentProxyList::_getComponents called at /usr/lib/perl/NCD/ComponentProxyList.pm line 777
 
                NCD::ComponentProxyList::_initialize called at /usr/lib/perl/CAF/Object.pm line 94
 
                CAF::Object::new called at /usr/sbin/ncm-ncd line 580
 
[ERROR] *** cannot instantiate class: NCD::ComponentProxy
 
                (in cleanup) finishing... at /usr/sbin/ncm-ncd line 425 during global destruction.
```

@ajf8 tracked this down to a bug in the perl  version module (shipped by Red Hat) which parses version strings. It used to be written in pure perl, but at some point they did a preferred C implementation, with fallback to the perl one. Which is why this works around it:
 
```# mv /usr/lib64/perl5/vendor_perl/auto/version/vxs/vxs.so{,.off}```
 
It’s somehow related to tainted data (which the version string is). I guess the new C implementation of version.pm is unable to handle tainted strings properly. [This will be reported to RH]
 
Turning off taint mode in the ncm-ncd executable works around the issue, but I think untainting the version using a regex in ComponentProxy.pm should be a better workaround.

--- a/src/main/perl/ComponentProxy.pm
+++ b/src/main/perl/ComponentProxy.pm
@@ -236,15 +236,22 @@ sub _initialize
     if ($tree->{version}) {
         # version is part of the PMpost maven template
         # be aware of obscure tainting errors (version uses XS); untaint if needed
-        local $@;
-        eval {
-            $self->{VERSION_CONFIG} = version->new($tree->{version});
-        };
-        if ($@) {
-            $self->error("component $name has invalid version from config $tree->{version} (bug in profile?): $@");
-            return;
+        my $errmsg = "component $name has invalid version from config $tree->{version} (bug in profile?)";
+        my $pattern = '^(v?\d+(?:\.\d+(?:\.\d+)))$';
+        if ($tree->{version} =~ m/$pattern/) {
+            local $@;
+            eval {
+                $self->{VERSION_CONFIG} = version->new($1);
+            };
+            if ($@) {
+                $self->error("$errmsg: $@");
+                return;
+            } else {
+                $self->verbose("component $name version from config $self->{VERSION_CONFIG}");
+            }
         } else {
-            $self->verbose("component $name version from config $self->{VERSION_CONFIG}");
+            $self->error("$errmsg: does not match regex pattern $pattern");
+            return;
         }
     } else {
         $self->verbose("component $name no version from config");

